### ISSUE TITLE
Crash when using singleton

### ISSUE THREAD

**Issue by totidev** [hmlongco/Factory#30]
I have found a SIGABRT error occurs under a specific scenario. In this case, I am trying to provide a protocol for UIApplication, and I am setting this as a singleton (this is so that I can provide a mock during testing). The issue appears to be caused when setting the scope as singleton, and the factory itself is a singleton. I am providing the code here which causes the issue. All you need to do is run the test. 
[FactoryDemo.zip](https://github.com/hmlongco/Factory/files/9784102/FactoryDemo.zip)

**Comment by hmlongco** [comment]
Don't do that?

It works if you don't do the singleton option, even though it shouldn't.

```swift
public class FactoryContainer: SharedContainer {
    public static let application = Factory<UIApplicationProtocol> {
        let app = UIApplication.shared
        return app
    }
}
```

If you breakpoint the return and run the test you'll see that app is (supposedly) uninitialized... but it still works. I added `isNetworkActivityIndicatorVisible: Bool` to the protocol and did the following.

```
(lldb) po app
 <uninitialized>
(lldb) po app.isNetworkActivityIndicatorVisible
false
(lldb) po UIApplication.shared
 <uninitialized>
(lldb) po UIApplication.shared.isNetworkActivityIndicatorVisible
false
```

Something weird is happening with the UIApplication instance.

Updated the code:

[FactoryDemo.zip](https://github.com/hmlongco/Factory/files/9787815/FactoryDemo.zip)

**Comment by hmlongco** [comment]
So I did some more research on this and obtained an answer in another context.

UIApplication.shared is declared to return UIApplication, so if you're accessing that value in a context where it's actually giving you a nil reference from the UIKit side, that's going to cause problems for Swift (because on the Swift side of things, we won't be performing any nil checks).

But UIApplication is also an Objective-C class, and so an access like UIApplication.shared.isNetworkActivityIndicatorVisible is going to become an Objective-C message sent to UIApplication.shared.

And it's perfectly valid in Objective-C to send messages to nil and get back 'reasonable' values for many return types...

Such as false for Bool values.

Bottom line is that UIApplication.shared is undefined at the time you're executing the Factory, and Swift is confused because  it was told that UIApplication.shared exists... when it does not.

**Comment by hmlongco** [comment]
So in your case if your injection is going to be resolved such that it could occur before the application actually exists, then you'll probably need to do...
```swift
public class FactoryContainer: SharedContainer {
    public static let application = Factory<UIApplicationProtocol?> { UIApplication.shared }
}
```
While optional may not be optimal, it may be your only choice.

**Comment by kamaldeep-earnin** [comment]
First of all I think it doesn't makes sense for a factory to have a `singleton` scope that returns a singleton instance. 

Secondly, the project is incomplete since its testing a UIKit API without a target depending based off `UIKit` framework. The application instance will be `nil` since the unit test target doesn't have a host application set.

<img width="415" alt="Screen Shot 2022-10-14 at 12 05 26 PM" src="https://user-images.githubusercontent.com/62305354/195922161-22c2a14d-9939-4490-9eab-a2c28f137128.png">

If we create a host target in there, and set it like below, the test would succeed since now since the test runner would be able to get an instance of `UIApplication`.

<img width="380" alt="Screen Shot 2022-10-14 at 12 09 21 PM" src="https://user-images.githubusercontent.com/62305354/195922859-6cad21a7-4aa3-4c5e-92aa-aa652cd174c7.png">

Updated project
[FactoryDemo-updated.zip](https://github.com/hmlongco/Factory/files/9789372/FactoryDemo-updated.zip)

**Comment by hmlongco** [comment]
Agree with the singleton of singletons, but the rest depends on when he's trying to instantiate his objects. Consider.
```swift
import SwiftUI
import Factory
import FactoryDemo

@main
struct HostApp: App {
    @Injected(FactoryContainer.application) var application
    var body: some Scene {
        WindowGroup {
            ContentView()
        }
    }
}
```
This will fail as HostApp (and its dependencies) are instantiated before the world exists. (i.e. .shared == nil)

**Comment by hmlongco** [comment]
This, however, works.
```swift
import SwiftUI
import FactoryDemo

@main
struct HostApp: App {
    @StateObject var example = ExampleInjected()
    var body: some Scene {
        WindowGroup {
            ContentView()
        }
    }
}
```
I tweaked the Example code a bit to make it an ObservableObject and added a public initializer, But ExampleInjected is thunked and not created until it's needed, which means by that point in time .shared is no longer nil.

**Comment by totidev** [comment]
When I created this sample, I had left out some key details - like why you might do something like I did. The reason is for testing purposes, where you might want to mock the UIApplication. The reason for using a singleton was to ensure that the same Mock instance was returned, rather than a new one. An example of this is if my mock had a Spy, and I wanted to ensure something was called. Using the default scope would not have worked as a new instance would have been returned.

One approach to this problem is just to keep the `application` using the default scope, and when I inject my mock, just to declare it locally, which will ensure the same instance is returned to the test.

**Comment by totidev** [comment]
> First of all I think it doesn't makes sense for a factory to have a `singleton` scope that returns a singleton instance.
> 
> Secondly, the project is incomplete since its testing a UIKit API without a target depending based off `UIKit` framework. The application instance will be `nil` since the unit test target doesn't have a host application set.
> 
> <img alt="Screen Shot 2022-10-14 at 12 05 26 PM" width="415" src="https://user-images.githubusercontent.com/62305354/195922161-22c2a14d-9939-4490-9eab-a2c28f137128.png">
> 
> If we create a host target in there, and set it like below, the test would succeed since now the test runner would be able to get an instance of `UIApplication`.
> 
> <img alt="Screen Shot 2022-10-14 at 12 09 21 PM" width="380" src="https://user-images.githubusercontent.com/62305354/195922859-6cad21a7-4aa3-4c5e-92aa-aa652cd174c7.png">
> 
> Updated project [FactoryDemo-updated.zip](https://github.com/hmlongco/Factory/files/9789372/FactoryDemo-updated.zip)

Thanks for this - I realise that this was an issue in my provided example. My tests should have been providing a mock for this, which was not the case here.

**Comment by hmlongco** [comment]
Nope. Understood the scenario, and have seen several instances of wanting to mock things like UIApplication.openURL().

Just need to be careful doing so since, as seen, with SwiftUI it's now possible to attempt to access UIApplication.shared before UIApplication even exists.

**Comment by hmlongco** [comment]
Hey, on reflection it's also possible to do... function injection!
```swift
typealias OpenURLFunction = (_ url: URL) -> Bool

extension Container {
    static let openURLFunction = Factory<OpenURLFunction> { UIApplication.shared.openURL }
}

struct OpenView: View {
    @Injected(Container.openURLFunction) var openURLFunction
    var body: some View {
        Button("Open") {
            _ = openURLFunction(URL(string: "https://www.google.com")!)
        }
    }
}
```

**Comment by totidev** [comment]
Thanks - that is useful to use, rather than having to create a protocol just to do one thing, eg: open urls from UIApplication.

**Comment by hmlongco** [comment]
Article on part of this

https://michaellong.medium.com/factory-and-functional-dependency-injection-2d0a38042d05
